### PR TITLE
feat: wasm module hot reload

### LINKED ISSUE #165: .wasm module hot reload
**What would you like to be added**:
Let Layotto monitor whether the .wasm file has changed, and reload the .wasm file if there is a change.
Achieve the effect of dynamically replacing .wasm at runtime

**Why is this needed**:
In order to implement subsequent functions:
1. FaaS based on wasm
2. Let developers change from developing sdk to developing wasm in the future, so that they can operate and maintain their wasm module independently

**How ​​to implement it**
1. After startup, monitor whether the specified .wasm file has changed. You can use some open source libraries, such as https://github.com/fsnotify/fsnotify
2. Once the file changed, call wasm related API to reload the file

**Reference**
[Multilingual programming based on WASM](https://mosn.io/layotto/#/en/start/wasm/start)
The WASM engine used by Layotto is [WASMER](https://github.com/wasmerio/wasmer-go) .You can refer to official documents and examples to use API

## chinese
**需要实现什么功能**
让Layotto能监听.wasm文件有没有变更，有变更的话重新热加载.wasm文件
达到运行时动态替换.wasm的效果

**做这个功能的价值**
为了实现后续功能：
1. 基于wasm做FaaS
2. 让开发者以后从开发sdk改成开发wasm，能够独立运维.见 #166

**大致方案**
1. 启动后监听指定.wasm文件有没有变更。可以使用一些开源库，例如https://github.com/fsnotify/fsnotify
2. 监听到变更后，调wasm相关API重新加载该文件

**Reference**
[使用WASM进行多语言编程](https://mosn.io/layotto/#/zh/start/wasm/start)
Layotto使用的WASM引擎是[WASMER](https://github.com/wasmerio/wasmer-go) 可以参考官方文档和example使用API

### PR DESCRIPTION
**Pull request by zu1k**
**What this PR does**:

Let Layotto monitor whether the .wasm file has changed, and reload the .wasm file if there is a change.
Achieve the effect of dynamically replacing .wasm at runtime

**Which issue(s) this PR fixes**:

Fixes #165

**Special notes for your reviewer**:

I'm using https://github.com/fsnotify/fsnotify to monitor whether the specified .wasm file has changed. However I found events may be lost in some cases for the principle of fsnotify. The fsnotify is based on the file system and OS features. For example, on GNU/Linux, inotify is taking the info from the `inode` of the file, and not the file name. If a file is deleted and it is created again, it may have a different inode. I'm trying to watch its parent directory to deal with this case.

When using `mv` to overwrite the wasm file, it will `remove` the file first, then `rename` the origin file. In this case, fsnotify will lost the monitor when got `remove` event, resulting in subsequent failure to continue hot loading. We won't have this problem with `cp`.  I'm trying to solve this problem by rewatching the file when the event is `remove` and file still exists. 

What's more, I found that mosn encapsulates wasmer but doesn't provide the ability to reload a wasm module, so currently I reload by uninstalling first and then adding, I don't think this is a good way to do it, if possible I would like someone to  expose the method by create a PR for mosn.

**Does this PR introduce a user-facing change?**:

NONE

### CODE REVIEW

**Comment by zu1k** [comment]
I'm trying to create a PR for `mosn` to provide the ability to reload a wasm plugin. See https://github.com/mosn/mosn/pull/1741 for more details.

If the PR merged, we can use a more convenient method to achieve wasm reload, see https://github.com/zu1k/layotto/commit/e26d11b2fdd4484c60a4782a8fe54dd57468b8a7 for the latest changes.

File: pkg/wasm/watcher.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+			if err != nil {
+				log.DefaultLogger.Errorf("[proxywasm] [watcher] reloadWasm fail to uninstall plugin, err: %v", err)
+			}
+
+			v2Config := v2.WasmPluginConfig{
+				PluginName:  pluginName,
+				VmConfig:    config.VmConfig,
+				InstanceNum: config.InstanceNum,
+			}
+			err = wasm.GetWasmManager().AddOrUpdateWasm(v2Config)
```
Comment (zhenjunMa) [inline pkg/wasm/watcher.go:102]: 这里应该只需要调这个AddOrUpdateWasm就可以吧？  上面卸载plugin跟下面重新创建应该不需要，配置什么的都没变

Comment (zu1k) [reply pkg/wasm/watcher.go:102]: 有多处检查保证了相同的配置直接返回错误

https://github.com/mosn/mosn/blob/a5104df830d357133c20a2b023d697cd434a8280/pkg/wasm/manager.go#L67
https://github.com/mosn/mosn/blob/a5104df830d357133c20a2b023d697cd434a8280/pkg/wasm/manager.go#L74

**Comment by seeflood** [comment]
Cool,thanks for your contribution!
I did not expect the fsnotify library to have such flaws...

File: pkg/wasm/watcher.go (hunk 1 above)
Comment (zhenjunMa) [reply pkg/wasm/watcher.go:102]: 非常抱歉这么久才回复，我这里只剩下一个问题，就是监听的文件如`a.wasm`如果被覆盖，则会触发两次加载。
因为覆盖操作会产生`remove`跟`create`两个事件，而这两个事件中`fileExist`都返回true。

复现流程：
准备两个wasm文件，`a.wasm`跟`b.wasm`，指定layotto启动的时候加载`a.wasm`，然后执行`mv  b.wasm a.wasm`即可。

Comment (zu1k) [reply pkg/wasm/watcher.go:102]: 如果你测试`cp`命令，你会得到多个`write``事件，这更会导致多次reload。

我认为这个行为符合预期，因为我们无法判断在哪一次write系统调用后文件是真正修改好的，所以只能在每一次write事件发生后都reload一下，以保证在文件真正被修改好后能够被reload上。

对于`mv`命令，strace得到的系统调用序列是:

```
renameat2(AT_FDCWD, "bb.wasm", AT_FDCWD, "wasm.wasm", RENAME_NOREPLACE) = -1 EEXIST (File exists)
stat("wasm.wasm", {st_mode=S_IFREG|0775, st_size=256217, ...}) = 0
lstat("bb.wasm", {st_mode=S_IFREG|0775, st_size=256217, ...}) = 0
newfstatat(AT_FDCWD, "wasm.wasm", {st_mode=S_IFREG|0775, st_size=256217, ...}, AT_SYMLINK_NOFOLLOW) = 0
stat("wasm.wasm", {st_mode=S_IFREG|0775, st_size=256217, ...}) = 0
access("wasm.wasm", W_OK)               = 0
rename("bb.wasm", "wasm.wasm")          = 0
```

可以看到`mv`进行了以下动作：
1. 使用renameat2并且设置RENAME_NOREPLACE标志位，因目标文件存在返回错误
2. stat\lstat后进行了newfstatat，然后重新stat
3. 使用access检查目标文件允许被写入
4. 使用rename直接将源文件重命名，操作系统直接进行了replace

在fsnotify中得到的event是：

```
18:32:04,415 [watcher] runWatcher got event, "demo/wasm/code/golang/bb.wasm": RENAME
18:32:04,415 [watcher] runWatcher got event, "demo/wasm/code/golang/wasm.wasm": CREATE
18:32:04,459 [watcher] runWatcher got event, "demo/wasm/code/golang/wasm.wasm": CHMOD
18:32:04,460 [watcher] runWatcher got event, "demo/wasm/code/golang/wasm.wasm": REMOVE
```

我暂时不清楚fsnotify是如何进行对应的，我会对mv的情况进行处理，但无法保证对多次write能够只reload一次

Comment (zhenjunMa) [reply pkg/wasm/watcher.go:105]: 棒！

**Review by zhenjunMa — APPROVED** [review]

**Comment by zhenjunMa** [comment]
@zu1k   thank you for your contribute!

**Pull request merged**